### DISCUSSION TITLE
sFlow data traffic issue

### DISCUSSION THREAD

**Discussion by MarianRychtecky** [akvorado/akvorado#719] [Q&A]
I don't want to open a BUG before I'm sure I have all configured properly. Can you please have a look and suggest what could be wrong? Many Thanks

```
# akvorado version
 ghcr.io/akvorado/akvorado:1.8.3
```

# Bug description
The data traffic volume in the graphs is much higher than SNMP readings.

Akvorado graphs show 194.24 Gbit/s in and 408.25 Gbit/s out of the port.
Akvorado graphs show 30.61 Mpps in and 42.48 Mpps out of the port.

SNMP graphs show 126 Gbit/s in and 252 Gbit/s out of the port.
SNMP shows 20 Mpps in and 26.88 Mpps out  of the port
SNMP shows 740.39 bytes average packet size in and 1.18 kbyte average packet size out of the port

Switch readings:
  Load-Interval #1: 30 seconds
    30 seconds input rate 127930730328 bits/sec, 20250179 packets/sec
    30 seconds output rate 256772244992 bits/sec, 27302812 packets/sec
    input rate 127.93 Gbps, 20.25 Mpps; output rate 256.77 Gbps, 27.30 Mpps

It looks like PPS is not calculated correctly and in some cases, it is double than real traffic.

## Steps to reproduce the problem

Compare graphs and show output from the switch

## Expected outcome

input rate 127.93 Gbps, 20.25 Mpps; output rate 256.77 Gbps, 27.30 Mpps
![2023-05-19_19-06-04](https://github.com/akvorado/akvorado/assets/36925394/98b90043-7c48-4a60-9172-6c00093e583c)
![2023-05-19_19-48-12](https://github.com/akvorado/akvorado/assets/36925394/b8ce0f4f-c8fd-41dc-b05d-efb4bcd2f290)
![2023-05-19_19-48-37](https://github.com/akvorado/akvorado/assets/36925394/bfe1beea-0082-43ea-ae5f-d33419d73317)

## Current outcome

Akvorado graphs show 194.24 Gbit/s in and 408.25 Gbit/s out of the port.
Akvorado graphs show 30.61 Mpps in and 42.48 Mpps out of the port.

![2023-05-19_19-49-14](https://github.com/akvorado/akvorado/assets/36925394/08ff986b-ad4c-47b6-a7a3-875212c9b261)
![2023-05-19_19-49-42](https://github.com/akvorado/akvorado/assets/36925394/324a7b20-eada-4c13-b964-96b92f0e8bde)

**Comment by vincentbernat** [comment]
Which equipment?

**Comment by MarianRychtecky** [comment]
Cisco Nexus 93600CD-GX, interface port-channel, 2x 400G ports in channel

**Reply by vincentbernat** [reply]
Sorry, no clue. You may want to restrict it to one interface, preferably a physical one. What sampling rate do you use?

**Comment by MarianRychtecky** [comment]
The difference is the same on the physical (single) interface. The sampling rate is 1:65536, and Akvorado correctly recognizes it.

**Reply by vincentbernat** [reply]
Maybe a packet capture may detect something odd. Could you share one sFlow packet?

**Reply by MarianRychtecky** [reply]
Thank you, I sent you a private email with my findings.

**Reply by vincentbernat** [reply]
I don't see anything suspicious. Can you check sflow counters from Akvorado? If you check the number of ExpandedFlowSample over a minute when sampling over a single interface and divide by the sampling rate, do you get the either the same number of PPS that is reported by the interface?

**Reply by MarianRychtecky** [reply]
I hope I understood correctly.

- Interval was 120s
- Only one physical interface is configured to send samples
- Sampling default - 4096
- Cisco reported 
```
  Load-Interval #1: 30 seconds
    30 seconds input rate 955544936 bits/sec, 352141 packets/sec
    30 seconds output rate 4466627496 bits/sec, 449417 packets/sec
    input rate 955.54 Mbps, 352.14 Kpps; output rate 4.47 Gbps, 449.42 Kpps
```

`
akvorado_inlet_flow_decoder_sflow_sample_sum{agent="",exporter="",type="ExpandedFlowSample",version="5"} 1.48162732e+08

....120s sleep ...

akvorado_inlet_flow_decoder_sflow_sample_sum{agent="",exporter="",type="ExpandedFlowSample",version="5"} 1.48186344e+08
`
Counter difference 23612

23612 / 120 = 196.766 flows/s
196.766 * 4096 = 805.956 pps

This looks to me that ExpandedFlowSample is PPS in both directions combined (449.42+352.12), but in GUI, this is reported in both directions.

![image](https://github.com/akvorado/akvorado/assets/36925394/e100ef14-3319-4ea8-8633-acd210aa07e3)

**Reply by vincentbernat** [reply]
For the latest capture, what's your filter? You should always include a direction in the filter (like `InIfBoundary = external` or `InIfName = ...`).

**Reply by MarianRychtecky** [reply]
Sorry, I didn't notice the filter disappear after the page refresh.

The difference on this switch (with smaller traffic and sampling rate configured at 1:4096) is about 12%. I will test it on the switch with higher traffic to see whether the difference will be larger.

Here is the corrected filter/graph.

![image](https://github.com/akvorado/akvorado/assets/36925394/f7f99067-a0d6-4522-bb3b-4a1775ab4cb6)

![image](https://github.com/akvorado/akvorado/assets/36925394/5e6aa08a-89a8-4379-b945-c7809bdee23f)

**Reply by vincentbernat** [reply]
12% on PPS or 12% on BPS? Note that L2BPS is also an approximation. Akvorado only keeps the L3 payload and deduce the L2 payload from that without taking VLAN into account and assuming a minimum overhead (it adds 38 bytes to the L3 packet size: MAC + ethernet type + FCS + preamble + IPG). It's a bit difficult to know what each switch displays for L2 speed.

On PPS, you should get the same result. Then, you can try to increase to 65536 to check you still get the same result, then add more physical interfaces to sampling, then add some logical interfaces (like a bundle).

**Reply by MarianRychtecky** [reply]
Thank you for your time, Vincent. I'll do more research and tests and I'll be back to you in couple of days.
